### ISSUE TITLE
sessionSampleRate and error reporting

### ISSUE THREAD

**Issue by renataogarcia** [aws-observability/aws-rum-web#472]
### What is your general inquiry?

I'm wondering if `sessionSampleRate` affects error reporting. Are errors always reported, or only for those sessions that are being sampled?

**Comment by williazz** [comment]
All HTTP errors are recorded. For HTTP requests, SessionSampleRate only affects traces

**Comment by williazz** [comment]
sessionSampleRate affects all RUM events

**Comment by renataogarcia** [comment]
@williazz  Thanks! Would make sense to create two sessions/instances with different sample rates, if I want to log all errors but record performance at a different rate?

**Comment by adebayor123** [comment]
Session sample rate is determined by the web client configuration. As a result, it's difficult to create two sessions with different sample rates, because that would require that you deploy two different versions of your application.

That being said, if you want to record all errors but record other data at a different rate, you can experiment initializing two web clients, one with `sessionSampleRate = 1` to capture all errors with only the Error plugin initialized, and the other with your custom sample rate and other plugins initialized.

Having two web clients initialized had some unintended consequences in the past (such as adding duplicate X-Ray traces), but it's always better to create a prototype and test it out.

**Comment by renataogarcia** [comment]
@adebayor123 Sorry, I meant two web clients. Ok, I'll try it out.

Also, as an alternative, would it be viable for me to create a custom or patched plugin that logs performance at different sample rate? or this would interfere with the stats reporting in the RUM dashboard?

**Comment by williazz** [comment]
> Also, as an alternative, would it be viable for me to create a custom or patched plugin that logs performance at different sample rate? or this would interfere with the stats reporting in the RUM dashboard?

No, that would be totally fine to implement. Just make sure to disable that plugin you want to ignore, and then refactor our plugin with a different sample rate.

**Comment by williazz** [comment]
Closing this issue, but feel free to reopen if you have issues implementing.

**Comment by ericxinzhang** [comment]
> Session sample rate is determined by the web client configuration. As a result, it's difficult to create two sessions with different sample rates, because that would require that you deploy two different versions of your application.
> 
> That being said, if you want to record all errors but record other data at a different rate, you can experiment initializing two web clients, one with `sessionSampleRate = 1` to capture all errors with only the Error plugin initialized, and the other with your custom sample rate and other plugins initialized.
> 
> Having two web clients initialized had some unintended consequences in the past (such as adding duplicate X-Ray traces), but it's always better to create a prototype and test it out.

I did a POC and find out that the two web clients approach won't work, as the two clients share the same cookie and will [try to restore the session from cookie](https://github.com/aws-observability/aws-rum-web/blob/ff5fda453e43503400b5cf2e4c580c3e8588d220/src/sessions/SessionManager.ts#L105-L106) so eventually the two clients have [the same `record` property value](https://github.com/aws-observability/aws-rum-web/blob/ff5fda453e43503400b5cf2e4c580c3e8588d220/src/sessions/SessionManager.ts#L92).

**Comment by qhanam** [comment]
Can we add a configuration option that provides customizable control over sampling? Using multiple instances feels hacky, because this would yield multiple session Ids for the same session.

For example:
1. Adding a configuration to the errors telemetry to set the sampling policy (e.g. `recordAll: true` ).
2. Adding a sampling hook/filter where the user provides a function to make the sampling decision (e.g., `samplingFilter: (event) -> event.type === 'com.amazonaws.error' ? true : false`).
